design/open/hard-delete.md

```diff
@@ -0,0 +1,307 @@
+      2. If state == 'deleted' in all entries and min(`last_modified`) < `TBD`
+         1. Hard-delete object
+         2. Delete all keys for `physical_address`
+
+### Key Update Order of Precedence
+
+1. `commited` state takes precedence over all (terminal state - overrides any other value) - uses **Set**
+2. `deleted` state can only be done on entries which are not `committed` and uses **SetIf**
+3. `staged` state can only be done on entries which are not `committed` and uses **SetIf** 
+
+### Flows
+
+#### Upload Object
+
+1. Write blob
+2. Write reference entry to the database as staged
+3. Add staged entry to database
+4. If entry exists in current staging token - mark old reference as deleted
+**Open:** Efficiently deal with overrides 
+
+#### Stage Object
+
+1. Object path is in the repo namespace:
+   1. Add reference entry to the database 
+2. Object path is outside the repo namespace:
+   1. lakeFS will not handle its retention
+3. Add staged entry
+
+#### GetPhysicalAddress
+Get physical address will add a reference to the generated unique physical address with state `deleted`
+We will then return a valid token / expiry timestamp to the user in addition to the physical address
+The user will need to pass the token to LinkPhysicalAddress
+
+#### LinkPhysicalAddress
+Add a validation to check if the token provided / or timestamp expired.
+Set reference state as staged.
+
+Assume this API uses a physical address in the repo namespace
+1. Add reference entry to the database
+2. Add staging entry
+
+#### Delete Object
+
+1. If object staged
+   1. Read reference key
+   2. If not `committed` 
+      1. Change reference state to `deleted` and update `last_modified`
+
+#### Commit
+
+1. Mark all entries in staging area as `committed`
+2. Swap staging token
+3. Create commit
+4. Update branch commit ID
+
+#### Reset / Delete Branch
+
+When resetting a branch we throw all the uncommitted branch data, this operation happens asynchronously via `AsyncDrop`.
+We can leverage this asynchronous job to also check and perform the hard-delete
+1. Scan deleted staging token entries
+2. If entry is `tombstone` remove entry reference key
+3. Otherwise, modify reference state to `deleted` and update `last_modified`
+
+## 4. Staging Token States
+
+Track state on the staging token instead of objects - but keep tracking references (without state)
+References will be tracked in a list instead of an explicit entry per reference
+    
+    key    = <repo>/references/<physical_address>  
+    value  = list(staging tokens)
+
+For each staging token of a deleted / reset branches a deleted entry will be added
+
+    <repo>/staging/deleted/<staging_token>
+
+For each commit we will add the staging tokens to a committed entry.
+Adding a reference is done using `set if` and a retry mechanism to ensure consistency.
+
+    key    = <repo>/references/<physical_address>  
+    value  = list(staging tokens)
```

Comment:
Looks like the same as for reference

Reply:
Sorry, there was a typo here - it refers to the `staging/committed' key. Fixed